File: java/LinkedList/Day9/IntersectionOfLL.java
package Day9;
import java.util.*;

class Node {
    int data;
    Node next;
    Node back;

    Node(int data1, Node next1, Node back1) {
        this.data = data1;
        this.next = next1;
        this.back = back1;
    }

    Node(int data1) {
        this.data = data1;
        this.next = null;
        this.back = null;
    }
}



public class IntersectionOfLL {
    private static Node intersectionLL(Node head1, Node head2) {

        // brute approach...
        HashSet<Node> set = new HashSet<>();
        Node temp = head1;

        while (temp != null) {
            set.add(temp);
            temp = temp.next;
        }

        temp = head2;

        while (temp != null) {
            if (set.contains(temp)) {
                return temp;
            }
            temp = temp.next;
        }

        return null;

    }





    




    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // -------- INPUT FIRST LIST --------
        System.out.print("Enter number of nodes in list 1: ");
        int n1 = sc.nextInt();

        Node head1 = null, tail1 = null;

        System.out.print("Enter elements of list 1: ");
        for (int i = 0; i < n1; i++) {
            int val = sc.nextInt();
            Node newNode = new Node(val);
            if (head1 == null) {
                head1 = newNode;
                tail1 = newNode;
            } else {
                tail1.next = newNode;
                tail1 = newNode;
            }
        }

        // -------- INPUT SECOND LIST --------
        System.out.print("Enter number of nodes in list 2: ");
        int n2 = sc.nextInt();

        Node head2 = null, tail2 = null;

        System.out.print("Enter elements of list 2: ");
        for (int i = 0; i < n2; i++) {
            int val = sc.nextInt();
            Node newNode = new Node(val);
            if (head2 == null) {
                head2 = newNode;
                tail2 = newNode;
            } else {
                tail2.next = newNode;
                tail2 = newNode;
            }
        }

        // -------- CONNECT INTERSECTION --------
        System.out.print("Enter position to intersect (1-based, 0 for no intersection): ");
        int pos = sc.nextInt();

        if (pos > 0) {
            Node temp = head1;
            for (int i = 1; i < pos; i++) {
                temp = temp.next;
            }
            tail2.next = temp;  // create intersection
        }

        // -------- FIND INTERSECTION --------
        Node result = intersectionLL(head1, head2);

        if (result != null)
            System.out.println("Intersection Node Value: " + result.data);
        else
            System.out.println("No intersection found");

        sc.close();
    }
}

File: python/LinkedList/day9/intersectionOfLL.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.back = None



def getIntersection(head1, head2):
    # brute approach...
    visited = set()
    temp = head1
    while temp is not None:
        visited.add(temp)
        temp = temp.next

    temp = head2

    while temp is not None:
        if temp in visited:
            return temp
        
        temp = temp.next
        
    return None








# ---------- MAIN ----------
n1 = int(input("Enter number of nodes in list 1: "))
print("Enter elements of list 1:", end=" ")
vals1 = list(map(int, input().split()))

head1 = None
prev = None
for v in vals1:
    node = Node(v)
    if head1 is None:
        head1 = node
    else:
        prev.next = node
    prev = node

n2 = int(input("Enter number of nodes in list 2: ",))
print("Enter elements of list 2:", end=" ")
vals2 = list(map(int, input().split()))

head2 = None
prev2 = None
for v in vals2:
    node = Node(v)
    if head2 is None:
        head2 = node
    else:
        prev2.next = node
    prev2 = node

# Connect intersection
pos = int(input("Enter intersection position (1-based, 0 for no intersection): "))
if pos > 0:
    temp = head1
    for _ in range(pos - 1):
        temp = temp.next
    prev2.next = temp

# Find intersection
ans = getIntersection(head1, head2)

if ans:
    print("Intersection Node:", ans.data)
else:
    print("No intersection found")